### PR TITLE
Remove torch cuda empty cache

### PR DESCRIPTION
**Pull request by shurkat-nvidia**
- [x] I, the PR author, have personally reviewed every line of this PR. (I have written every line of this PR.)

# What does this PR do?

The PR removes an expensive `torch.cuda.empty_cache()` call within `save_checkpoint_and_time()`. This call took multiple seconds on every checkpoint and also increased the size of the `generate_state_dict()` function by multiple seconds. To avoid OOM errors, the code also modifies the Nemotron3 Super script to 1) use SHM to avoid expensive `rebuild_cuda_tensor()` calls during the IPC handoff and 2) garbage collect the torch allocator at 80% GPU allocation with a lower overhead than a raw `empty_cache()` call, and to use larger blocks to prevent fragmentation in the first place.

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [ ] (N/A) I have added relevant unit tests
- [ ] (N/A) I have added relevant functional tests
- [ ] (N/A) I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] (N/A) I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

**Comment by asolergi-nv** [comment]
/ok to test [debb98a](https://github.com/NVIDIA/Megatron-LM/pull/5678/commits/debb98a11c0283b7cd425bf9e6e44e3f4def6556)

**Review by cuichenx — APPROVED** [review]

**Comment by cuichenx** [comment]
This is a replay of the change reverted in #5366, is CI passing with this PR?

**Comment by Phlip79** [comment]
> This is a replay of the change reverted in #5366, is CI passing with this PR?

This test was recently updated in https://github.com/NVIDIA/Megatron-LM/pull/6239. @shurkat-nvidia can you please check to see if the changes to the config are needed anymore?

**Comment by shurkat-nvidia** [comment]
> This is a replay of the change reverted in https://github.com/NVIDIA/Megatron-LM/pull/5366, is CI passing with this PR?

@cuichenx, @balasaajay ran the test manually a few weeks ago, it appeared to have addressed the OOM issue.

> This test was recently updated in https://github.com/NVIDIA/Megatron-LM/pull/6239. @shurkat-nvidia can you please check to see if the changes to the config are needed anymore?

@Phlip79, I rebased the branch to the latest main. Still need changes to checkpoint args and to the `PYTORCH_ALLOC_CONF` environment variable.

File: tests/functional_tests/test_cases/nemotron/nemotron3_super_release_gb200/model_config.yaml [hunk 1]
```diff
@@ -13,6 +13,7 @@ ENV_VARS:
   NVTE_FWD_LAYERNORM_SM_MARGIN: 20
   TORCHINDUCTOR_WORKER_START: fork
   QUANTIZATION_TYPE_DEBUG: 1
+  PYTORCH_ALLOC_CONF: garbage_collection_threshold:0.8,roundup_power2_divisions:true,max_non_split_rounding_mb:1024
```
Comment (deepakn94) [inline tests/functional_tests/test_cases/nemotron/nemotron3_super_release_gb200/model_config.yaml]: Strip to `garbage_collection_threshold:0.8` if safe.

Comment (shurkat-nvidia) [reply tests/functional_tests/test_cases/nemotron/nemotron3_super_release_gb200/model_config.yaml]: Done. Merged `PYTORCH_CUDA_ALLOC_CONF` setting.

**Review by deepakn94 — APPROVED** [review]

**Comment by deepakn94** [comment]
/ok to test [0f9153b](https://github.com/NVIDIA/Megatron-LM/pull/5678/commits/0f9153bfff72fb42fcc760a64d82021cca40ba30)
